mlir/lib/Catalyst/Transforms/BufferizableOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,284 @@
+#include "mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Bufferization/IR/UnstructuredControlFlow.h"
+#include "mlir/Dialect/Bufferization/Transforms/Bufferize.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Catalyst/IR/CatalystOps.h"
+#include "Catalyst/Transforms/BufferizableOpInterfaceImpl.h"
+
+using namespace mlir;
+using namespace catalyst;
+
+namespace {
+
+/// Bufferization of catalyst.print. Get memref of printOp.val.
+struct PrintOpInterface
+    : public bufferization::BufferizableOpInterface::ExternalModel<PrintOpInterface, PrintOp> {
+    bool bufferizesToMemoryRead(Operation *op, OpOperand &opOperand,
+                                const bufferization::AnalysisState &state) const
+    {
+        return true;
+    }
+
+    bool bufferizesToMemoryWrite(Operation *op, OpOperand &opOperand,
+                                 const bufferization::AnalysisState &state) const
+    {
+        return false;
+    }
+
+    bufferization::AliasingValueList
+    getAliasingValues(Operation *op, OpOperand &opOperand,
+                      const bufferization::AnalysisState &state) const
+    {
+        return {};
+    }
+
+    LogicalResult bufferize(Operation *op, RewriterBase &rewriter,
+                            const bufferization::BufferizationOptions &options) const
+    {
+        auto printOp = cast<PrintOp>(op);
+        if (printOp.getVal()) {
+            FailureOr<Value> source = getBuffer(rewriter, printOp.getVal(), options);
+            if (failed(source))
+                return failure();
+            bufferization::replaceOpWithNewBufferizedOp<PrintOp>(
+                rewriter, op, *source, printOp.getConstValAttr(), printOp.getPrintDescriptorAttr());
+        }
+        return success();
+    }
+};
+
+/// Bufferization of catalyst.print. Mainly get buffers for arguments.
+struct CustomCallOpInterface
+    : public bufferization::BufferizableOpInterface::ExternalModel<CustomCallOpInterface,
+                                                                   CustomCallOp> {
+    bool bufferizesToMemoryRead(Operation *op, OpOperand &opOperand,
+                                const bufferization::AnalysisState &state) const
+    {
+        return true;
+    }
+
+    bool bufferizesToMemoryWrite(Operation *op, OpOperand &opOperand,
+                                 const bufferization::AnalysisState &state) const
+    {
+        return false;
```

Comment:
What do you think @erick-xanadu, any chance we can mandate that callback arrays are not modified in place or should be conservative here?

Reply:
>  Or maybe we even already make copies at the boundary ourselves?

I think that's what I'm trying to answer. We don't make a copy, but I believe JAX will refuse to write to that data buffer because it is not owned by it. We just provide a view to it.

Reply:
Ah, this is a `CustomCallOp` not a `CallbackCallOp`. I do not know about the invariants of this operation.

Reply:
Good point, I think I was actually thinking about `CallbackCallOp` as well, so if you have any insights into the above questions for that one it would still be useful. 
For the `CustomCallOp` the semantics are probably easier to determine. We should be able to double check whether the blas routines do anything in-place or not

Reply:
This can be safely kept as false. What I find out is that this is only for operands and not results. Results will be allocated in a new buffer. So every IR operation that is not in DPS style and returns something can be safely assumed to only read buffers.

Reply:
> So every IR operation that is not in DPS style and returns something can be safely assumed to only read buffers.

Are you sure about this statement? Imo an operation that lowers to arbitrary code (e.g. external functions) could easily write to the input buffers even if it is not in DPS style or doesn't have results.